File: scripts/build_social_preview.mjs
#!/usr/bin/env node
/** Render the test site's current-wordmark 1200 x 630 social-share image. */

import path from "node:path";
import { fileURLToPath, pathToFileURL } from "node:url";


const WIDTH = 1200;
const HEIGHT = 630;
const modulePath = process.env.KC_SHARP_MODULE || "sharp";
const sharpModule = modulePath.startsWith("/") ? pathToFileURL(modulePath).href : modulePath;
const { default: sharp } = await import(sharpModule);

const scriptDir = path.dirname(fileURLToPath(import.meta.url));
const root = path.resolve(scriptDir, "..");
const output = path.resolve(
  process.argv[2] || path.join(root, "test-overrides/assets/social-preview-wordmark-20260917.png"),
);

// Rasterize the exact SVG used in the visible site header, then trim its
// transparent viewBox padding so the current wordmark is truly centered.
const renderedWordmark = await sharp(path.join(root, "assets/logo-wordmark.svg"), { density: 300 })
  .png()
  .toBuffer();
const wordmark = await sharp(renderedWordmark)
  .trim()
  .resize({ width: 960 })
  .png()
  .toBuffer();
const wordmarkInfo = await sharp(wordmark).metadata();

const background = `<svg xmlns="http://www.w3.org/2000/svg" width="${WIDTH}" height="${HEIGHT}">
  <defs>
    <radialGradient id="glow" cx="88%" cy="30%" r="70%">
      <stop offset="0" stop-color="#c84a32" stop-opacity=".20"/>
      <stop offset=".42" stop-color="#c84a32" stop-opacity=".07"/>
      <stop offset="1" stop-color="#080808" stop-opacity="0"/>
    </radialGradient>
    <linearGradient id="accent" x1="0" x2="1">
      <stop stop-color="#c84a32"/>
      <stop offset="1" stop-color="#df725c"/>
    </linearGradient>
  </defs>
  <rect width="1200" height="630" fill="#080808"/>
  <rect width="1200" height="630" fill="url(#glow)"/>
  <rect x="34" y="34" width="1132" height="562" rx="20" fill="none" stroke="#3a211c" stroke-width="2"/>
  <rect x="96" y="76" width="1008" height="4" rx="2" fill="url(#accent)"/>
  <text x="600" y="420" text-anchor="middle" fill="#f5f2ea" font-family="Arial, Helvetica, sans-serif" font-size="49" font-weight="700">Find Christian Hip-Hop Shows &amp; Festivals</text>
  <line x1="100" y1="514" x2="390" y2="514" stroke="#8f3928" stroke-width="2"/>
  <line x1="810" y1="514" x2="1100" y2="514" stroke="#8f3928" stroke-width="2"/>
  <text x="600" y="524" text-anchor="middle" fill="#df725c" font-family="Arial, Helvetica, sans-serif" font-size="21" font-weight="700" letter-spacing="5">KINGDOMCIRCUIT.COM</text>
</svg>`;

await sharp(Buffer.from(background))
  .composite([
    {
      input: wordmark,
      left: Math.round((WIDTH - wordmarkInfo.width) / 2),
      top: 155,
    },
  ])
  .png({ compressionLevel: 9 })
  .toFile(output);

process.stdout.write(`${output}\n`);

File: scripts/build_social_preview.py
#!/usr/bin/env python3
"""Render the test site's branded 1200 x 630 social-share image."""

from __future__ import annotations

import argparse
from pathlib import Path

from PIL import Image, ImageDraw, ImageFont


WIDTH = 1200
HEIGHT = 630
BACKGROUND = "#050505"
OFF_WHITE = "#f4f3ef"
GOLD = "#c99a3d"


def bold_font_path() -> Path:
    candidates = (
        Path("/System/Library/Fonts/Supplemental/Arial Bold.ttf"),
        Path("/usr/share/fonts/truetype/dejavu/DejaVuSans-Bold.ttf"),
    )
    for candidate in candidates:
        if candidate.is_file():
            return candidate
    raise FileNotFoundError("A supported bold font was not found")


def fit_font(draw: ImageDraw.ImageDraw, text: str, font_path: Path, max_width: int) -> ImageFont.FreeTypeFont:
    for size in range(48, 25, -1):
        font = ImageFont.truetype(str(font_path), size=size)
        left, _, right, _ = draw.textbbox((0, 0), text, font=font)
        if right - left <= max_width:
            return font
    return ImageFont.truetype(str(font_path), size=26)


def tracked_width(draw: ImageDraw.ImageDraw, text: str, font: ImageFont.FreeTypeFont, spacing: int) -> int:
    widths = [draw.textlength(character, font=font) for character in text]
    return round(sum(widths) + spacing * max(0, len(text) - 1))


def draw_tracked_text(
    draw: ImageDraw.ImageDraw,
    position: tuple[int, int],
    text: str,
    font: ImageFont.FreeTypeFont,
    fill: str,
    spacing: int,
) -> None:
    x, y = position
    for character in text:
        draw.text((x, y), character, font=font, fill=fill)
        x += round(draw.textlength(character, font=font)) + spacing


def render(logo_path: Path, output_path: Path) -> None:
    canvas = Image.new("RGB", (WIDTH, HEIGHT), BACKGROUND)
    draw = ImageDraw.Draw(canvas)

    # Keep every important element well inside social-platform crop-safe margins.
    draw.rounded_rectangle((28, 28, WIDTH - 29, HEIGHT - 29), radius=18, outline="#24211b", width=2)
    draw.line((82, 54, WIDTH - 82, 54), fill=GOLD, width=3)

    logo = Image.open(logo_path).convert("RGB")
    logo.thumbnail((790, 421), Image.Resampling.LANCZOS)
    logo_x = (WIDTH - logo.width) // 2
    canvas.paste(logo, (logo_x, 66))

    font_path = bold_font_path()
    headline = "Find Christian Hip-Hop Shows & Festivals"
    headline_font = fit_font(draw, headline, font_path, 1010)
    headline_box = draw.textbbox((0, 0), headline, font=headline_font)
    headline_width = headline_box[2] - headline_box[0]
    draw.text(
        ((WIDTH - headline_width) // 2, 496),
        headline,
        font=headline_font,
        fill=OFF_WHITE,
    )

    site = "KINGDOMCIRCUIT.COM"
    site_font = ImageFont.truetype(str(font_path), size=19)
    site_spacing = 3
    site_width = tracked_width(draw, site, site_font, site_spacing)
    site_x = (WIDTH - site_width) // 2
    site_y = 560
    draw_tracked_text(draw, (site_x, site_y), site, site_font, GOLD, site_spacing)
    line_y = site_y + 12
    draw.line((84, line_y, site_x - 30, line_y), fill="#795d27", width=2)
    draw.line((site_x + site_width + 30, line_y, WIDTH - 84, line_y), fill="#795d27", width=2)

    output_path.parent.mkdir(parents=True, exist_ok=True)
    canvas.save(output_path, "PNG", optimize=True)


def main() -> None:
    root = Path(__file__).resolve().parents[1]
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument(
        "--output",
        type=Path,
        default=root / "test-overrides" / "assets" / "social-preview.png",
    )
    args = parser.parse_args()
    render(root / "assets" / "logo.png", args.output.resolve())


if __name__ == "__main__":
    main()
